File: csharp/nucleotide-count/Dna.cs
using System.Collections.Generic;

public class Dna
{
    public Dictionary<char, int> NucleotideCounts { get; }

    public Dna(string nucleotides)
    {
        NucleotideCounts = new Dictionary<char, int> { { 'A', 0 }, { 'T', 0 }, { 'C', 0 }, { 'G', 0 } };
        CalculateNucleotideCounts(nucleotides);
    }

    public int Count(char nucleotide)
    {
        if (!NucleotideCounts.ContainsKey(nucleotide))
        {
            throw new InvalidNucleotideException();
        }
        return NucleotideCounts[nucleotide];
    }

    private void CalculateNucleotideCounts(string nucleotides)
    {
        foreach (var nucleotide in nucleotides)
        {
            NucleotideCounts[nucleotide]++;
        }
    }
}

File: swift/nucleotide-count/Dna.swift
import Foundation

class DNA {
    var dictionary : [String:Int] = ["A": 0, "G": 0, "T": 0, "C": 0]
    
    init?(strand: String) {
        for (_, value) in strand.characters.enumerated() {
            let nucleotide = String(value)
            if dictionary[nucleotide] == nil {
                return nil
            }
            dictionary[nucleotide]! += 1
        }
    }
    
    func count(_ nucleotide : String) -> Int? {
        return dictionary[String(nucleotide)];
    }
    
    func counts() -> [String:Int] {
        return dictionary;
    }
}
